### PR TITLE
Make DenseLayout noise aware

### PR DESCRIPTION
**Pull request by nonhermitian**
### Summary
This makes the `DenseLayout` noise aware.  The cost of a sub-graph is just the number of `cx` gates in the circuit * the avg cx err in the sub-graph plus the number of measurements times the difference between the avg meas err on the sub-set of qubits vs the whole device.  This can likely be improved upon, but it gets the job done.

I made this the mapping choice for optim_level=2. 

`DenseLayout` also no longer computes min-bandwidth ordering for each sub-graph anymore for efficiency.

### CODE REVIEW

**Comment by 1ucian0** [comment]
It seems different enough to be its own pass. Maybe `noise_dense_layout` or  `noise_aware_dense_layout`?

**Comment by nonhermitian** [comment]
This is the exact same layout method, just picks a better sub-graph from the equivalent set given the noise properties.

**Comment by ajavadia** [comment]
if backend_prop does not exist then this will just behave as before right? if so this is a good approach. 

But for the trivial_layout let's not do noise-aware (as is proposed in #3102), because trivial layout has explicitly been defined as an `i->i` mapping of qubits and this is something many people want. What #3102 proposes can be a NoiseAwareLinearLayout or something, but i'm not sure how useful that is so I'd recommend not doing it before a need arises.

File: qiskit/transpiler/passes/mapping/dense_layout.py [hunk 1]
```diff
@@ -36,18 +36,24 @@ class DenseLayout(AnalysisPass):
...
 
         Raises:
             TranspilerError: if invalid options
         """
         super().__init__()
         self.coupling_map = coupling_map
+        self.backend_prop = backend_prop
+        self.cx_mat = None
+        self.meas_arr = None
+        self.num_cx = None
```
Comment (1ucian0) [inline qiskit/transpiler/passes/mapping/dense_layout.py]:
```suggestion
        self.num_cx = 0
```

File: qiskit/transpiler/passes/mapping/dense_layout.py [hunk 2]
```diff
@@ -36,18 +36,24 @@ class DenseLayout(AnalysisPass):
...
+        self.num_meas = None
```
Comment (1ucian0) [inline qiskit/transpiler/passes/mapping/dense_layout.py]: Similarly here.

File: qiskit/transpiler/passes/mapping/dense_layout.py (hunk 1 above)
Comment (nonhermitian) [reply qiskit/transpiler/passes/mapping/dense_layout.py]: done

File: qiskit/transpiler/passes/mapping/dense_layout.py (hunk 2 above)
Comment (nonhermitian) [reply qiskit/transpiler/passes/mapping/dense_layout.py]: done, thanks for catching that.

**Pull request merged**